Mathlib/Algebra/Module/LocalizedModule.lean

```diff
@@ -0,0 +1,1122 @@
+
+private theorem add_comm' (x y : LocalizedModule S M) : x + y = y + x :=
+  LocalizedModule.induction_on₂ (fun m m' s s' => by rw [mk_add_mk, mk_add_mk, add_comm, mul_comm])
+    x y
+-- #align localized_module.add_comm' localized_module.add_comm'
+
+private theorem zero_add' (x : LocalizedModule S M) : 0 + x = x :=
+  induction_on
+    (fun m s => by
+      rw [← zero_mk s, mk_add_mk, smul_zero, zero_add, mk_eq];
+        exact ⟨1, by rw [one_smul, mul_smul, one_smul]⟩)
+    x
+-- #align localized_module.zero_add' localized_module.zero_add'
+
+private theorem add_zero' (x : LocalizedModule S M) : x + 0 = x :=
+  induction_on
+    (fun m s => by
+      rw [← zero_mk s, mk_add_mk, smul_zero, add_zero, mk_eq];
+        exact ⟨1, by rw [one_smul, mul_smul, one_smul]⟩)
+    x
+-- #align localized_module.add_zero' localized_module.add_zero'
+
+instance hasNatSmul : SMul ℕ (LocalizedModule S M) where smul n := nsmulRec n
+#align localized_module.has_nat_smul LocalizedModule.hasNatSmul
+
+private theorem nsmul_zero' (x : LocalizedModule S M) : (0 : ℕ) • x = 0 :=
+  LocalizedModule.induction_on (fun _ _ => rfl) x
+-- #align localized_module.nsmul_zero' localized_module.nsmul_zero'
+
+private theorem nsmul_succ' (n : ℕ) (x : LocalizedModule S M) : n.succ • x = x + n • x :=
+  LocalizedModule.induction_on (fun _ _ => rfl) x
+-- #align localized_module.nsmul_succ' localized_module.nsmul_succ'
+
+instance : AddCommMonoid (LocalizedModule S M) where
+  add := (· + ·)
+  add_assoc := add_assoc'
+  zero := 0
+  zero_add := zero_add'
+  add_zero := add_zero'
+  nsmul := (· • ·)
+  nsmul_zero := nsmul_zero'
+  nsmul_succ := nsmul_succ'
+  add_comm := add_comm'
+
+-- Porting note: prepared for `add_left_neg'`
+instance {M : Type _} [AddCommGroup M] [Module R M] : Neg (LocalizedModule S M) :=
+  { show AddCommMonoid (LocalizedModule S M) by
+      infer_instance with
+    neg := fun p =>
+      liftOn p (fun x => LocalizedModule.mk (-x.1) x.2) fun ⟨m1, s1⟩ ⟨m2, s2⟩ ⟨u, hu⟩ => by
+        rw [mk_eq]
+        exact ⟨u, by simpa⟩ }
+
+-- Porting note: This is stated as the separated lemma since it will be used twice.
+private theorem add_left_neg' {M : Type _} [AddCommGroup M] [Module R M]
+    (x : LocalizedModule S M) : -x + x = 0 := by
+  rcases x with ⟨m, s⟩
+  change
+    (liftOn (mk m s) (fun x => mk (-x.1) x.2) fun ⟨m1, s1⟩ ⟨m2, s2⟩ ⟨u, hu⟩ => by
+          rw [mk_eq]
+          exact ⟨u, by simpa⟩) +
+        mk m s =
+      0
+  rw [liftOn_mk, mk_add_mk]
+  simp
+
+instance {M : Type _} [AddCommGroup M] [Module R M] : AddCommGroup (LocalizedModule S M) :=
+  { show AddCommMonoid (LocalizedModule S M) by infer_instance,
+    show Neg (LocalizedModule S M) by infer_instance with
+    add_left_neg := add_left_neg' }
+
+theorem mk_neg {M : Type _} [AddCommGroup M] [Module R M] {m : M} {s : S} : mk (-m) s = -mk m s :=
+  rfl
+#align localized_module.mk_neg LocalizedModule.mk_neg
+
+set_option maxHeartbeats 400000 in
```

Comment:
Can you try this post #3414? If that fixes it, please make this PR dependent on that.

Reply:
This seem to be needed even after merging #3414.